cmd/producer/main_test.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2020 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/go-redis/redis/v8"
+)
+
+type fakeRedis struct {
+	client redis.Cmdable
+}
+
+func TestAsyncRequestHeader(t *testing.T) {
+	testserver := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != "GET" && r.Method != "POST" {
+			t.Errorf("Expected 'POST' OR 'GET' request, got '%s'", r.Method)
+		}
+	}))
+
+	tests := []struct {
+		name             string
+		async            bool
+		method           string
+		body             string
+		contentLengthSet bool
+		returncode       int
+	}{{
+		name:       "async get request",
+		async:      true,
+		method:     "GET",
+		body:       "",
+		returncode: 202,
+	}, {
+		name:       "non async get request",
+		async:      false,
+		method:     "GET",
+		body:       "",
+		returncode: 200,
+	}, {
+		name:       "non async post request",
+		async:      false,
+		method:     "POST",
+		body:       `{"body":"this is a body"}`,
+		returncode: 200,
+	}, {
+		name:       "async post request with too large payload",
+		async:      true,
+		method:     "POST",
+		body:       `{"body":"this is a larger body"}`,
+		returncode: 500,
+	}, {
+		name:       "async post request with smaller than limit payload",
+		async:      true,
+		method:     "POST",
+		body:       `{"body":"this is a body"}`,
+		returncode: 202,
+	}, {
+		name:       "test failure to write to Redis",
+		async:      true,
+		method:     "POST",
+		body:       "failure",
+		returncode: 500,
+	},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			env = envInfo{
+				StreamName:       "mystream",
+				RedisAddress:     "address",
+				RequestSizeLimit: "25",
+			}
+			setupRedis()
```

Comment:
Is it possible to setup fake redis only once, instead of for each test case?

Reply:
yes - thanks!